File: Tyche.Android/pool/src/main/java/com/felipearpa/tyche/pool/ui/poolscore/PoolScoreListViewModel.kt
package com.felipearpa.tyche.pool.ui.poolscore

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import androidx.paging.Pager
import androidx.paging.PagingConfig
import androidx.paging.PagingData
import androidx.paging.cachedIn
import com.felipearpa.tyche.pool.application.GetPoolGamblerScoresByGamblerUseCase
import com.felipearpa.tyche.pool.ui.PoolGamblerScoreModel
import com.felipearpa.tyche.pool.ui.PoolGamblerScorePagingSource
import dagger.assisted.Assisted
import dagger.assisted.AssistedInject
import kotlinx.coroutines.flow.Flow

private const val PAGE_SIZE = 50
private const val PREFETCH_DISTANCE = 5

class PoolScoreListViewModel @AssistedInject constructor(
    @Assisted private val gamblerId: String,
    private val getPoolGamblerScoresByGamblerUseCase: GetPoolGamblerScoresByGamblerUseCase
) :
    ViewModel() {

    val pageSize = PAGE_SIZE

    private lateinit var poolGamblerScorePagingSource: PoolGamblerScorePagingSource

    private var _searchText: String? = null

    private var _poolGamblerScores = Pager(
        config = PagingConfig(
            pageSize = pageSize,
            prefetchDistance = PREFETCH_DISTANCE,
            enablePlaceholders = true
        ),
        pagingSourceFactory = {
            PoolGamblerScorePagingSource(
                pagingQuery = { next ->
                    getPoolGamblerScoresByGamblerPagingQuery(
                        next = next,
                        gamblerId = gamblerId,
                        search = { _searchText },
                        getPoolGamblerScoresByGamblerUseCase = getPoolGamblerScoresByGamblerUseCase
                    )
                }
            ).also { poolPagingSource ->
                this.poolGamblerScorePagingSource = poolPagingSource
            }
        }
    ).flow.cachedIn(scope = viewModelScope)
    val poolGamblerScores: Flow<PagingData<PoolGamblerScoreModel>>
        get() = _poolGamblerScores

    fun search(searchText: String) {
        _searchText = searchText.trim().ifEmpty { null }
        poolGamblerScorePagingSource.invalidate()
    }
}

File: Tyche.iOS/Pool/Sources/Pool/PoolScore/PoolScoreListViewModel.swift
import Foundation
import Core
import UI
import DataPool

public class PoolScoreListViewModel: ObservableObject {
    private let getPoolGamblerScoresByGamblerUseCase : GetPoolGamblerScoresByGamblerUseCase
    
    private let gamblerId: String
    
    private var searchText: String? = nil
    
    private var pagingSource: PoolGamblerScorePagingSource!
    
    lazy var lazyPager: LazyPager<String, PoolGamblerScoreModel> = {
        LazyPager(
            pagingData: PagingData(
                pagingConfig: PagingConfig(prefetchDistance: 5),
                pagingSourceFactory: self.pagingSource
            )
        )
    }()
    
    public init(
        getPoolGamblerScoresByGamblerUseCase : GetPoolGamblerScoresByGamblerUseCase,
        gamblerId: String
    ) {
        self.getPoolGamblerScoresByGamblerUseCase = getPoolGamblerScoresByGamblerUseCase
        self.gamblerId = gamblerId
        
        self.pagingSource = PoolGamblerScorePagingSource(
            pagingQuery: { [unowned self] next in
                await getPoolGamblerScoresByGamblerUseCase.execute(
                    gamblerId: self.gamblerId,
                    next: next,
                    searchText: self.searchText
                )
                .map { page in page.map { poolGamblerScore in poolGamblerScore.toPoolGamblerScoreModel() }}
                .mapError { error in error.toLocalizedError() }
            }
        )
    }
    
    func search(_ newSearchText: String) {
        searchText = newSearchText
        pagingSource.invalidate()
    }
}

private extension Error {
    func toLocalizedError() -> Error {
        if let networkError = self as? NetworkError {
            return networkError.toNetworkLocalizedError()
        }
        
        return UnknownLocalizedError()
    }
}
